File: backend/routes/achievements.py
"""
Achievement routes for VerveQ Platform API
Simple MVP implementation following CLAUDE.md principles (<300 lines)
"""
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func, and_
from typing import List, Dict, Any
from pydantic import BaseModel
from datetime import datetime

from database.connection import get_db
from database.models import Achievement, UserAchievement, User, GameSession, UserRating

router = APIRouter(prefix="/achievements", tags=["achievements"])

class AchievementResponse(BaseModel):
    id: str
    name: str
    description: str
    category: str
    icon: str
    points: int
    is_hidden: bool

class UserAchievementResponse(BaseModel):
    achievement: AchievementResponse
    unlocked_at: datetime
    progress: Dict[str, Any] = None

class AchievementCheckResponse(BaseModel):
    newly_unlocked: List[str]
    total_unlocked: int
    total_points: int

@router.get("/", response_model=List[AchievementResponse])
async def list_achievements(db: Session = Depends(get_db)):
    """List all available achievements"""
    achievements = db.query(Achievement).filter(Achievement.is_hidden == False).all()
    return [
        AchievementResponse(
            id=a.id,
            name=a.name,
            description=a.description,
            category=a.category,
            icon=a.icon,
            points=a.points,
            is_hidden=a.is_hidden
        )
        for a in achievements
    ]

@router.get("/user/{user_id}", response_model=List[UserAchievementResponse])
async def get_user_achievements(user_id: str, db: Session = Depends(get_db)):
    """Get user's unlocked achievements"""
    # Verify user exists
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
    
    user_achievements = db.query(UserAchievement, Achievement).join(
        Achievement, UserAchievement.achievement_id == Achievement.id
    ).filter(UserAchievement.user_id == user_id).all()
    
    return [
        UserAchievementResponse(
            achievement=AchievementResponse(
                id=achievement.id,
                name=achievement.name,
                description=achievement.description,
                category=achievement.category,
                icon=achievement.icon,
                points=achievement.points,
                is_hidden=achievement.is_hidden
            ),
            unlocked_at=user_achievement.unlocked_at,
            progress=user_achievement.progress
        )
        for user_achievement, achievement in user_achievements
    ]

@router.post("/check/{user_id}", response_model=AchievementCheckResponse)
async def check_achievements(user_id: str, db: Session = Depends(get_db)):
    """Check and unlock new achievements for user (MVP implementation)"""
    # Verify user exists
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
    
    # Get already unlocked achievements
    unlocked_achievement_ids = set(
        ua.achievement_id for ua in 
        db.query(UserAchievement).filter(UserAchievement.user_id == user_id).all()
    )
    
    # Get user stats for checking achievements
    total_games = db.query(GameSession).filter(GameSession.user_id == user_id).count()
    quiz_games = db.query(GameSession).filter(
        and_(GameSession.user_id == user_id, GameSession.mode == 'quiz')
    ).count()
    survival_games = db.query(GameSession).filter(
        and_(GameSession.user_id == user_id, GameSession.mode == 'survival')
    ).count()
    
    # Get highest scores
    max_quiz_accuracy = db.query(func.max(GameSession.accuracy)).filter(
        and_(GameSession.user_id == user_id, GameSession.mode == 'quiz')
    ).scalar() or 0
    
    max_survival_score = db.query(func.max(GameSession.score)).filter(
        and_(GameSession.user_id == user_id, GameSession.mode == 'survival')
    ).scalar() or 0
    
    # Get sports played
    sports_played = db.query(GameSession.sport).filter(
        GameSession.user_id == user_id
    ).distinct().count()
    
    # Get highest ELO rating
    max_elo = db.query(func.max(UserRating.elo_rating)).filter(
        UserRating.user_id == user_id
    ).scalar() or 1200
    
    newly_unlocked = []
    
    # Check each achievement (simple MVP logic)
    achievement_checks = [
        ("first_quiz", quiz_games >= 1),
        ("first_survival", survival_games >= 1),
        ("quiz_master", max_quiz_accuracy >= 100),
        ("survival_legend", max_survival_score >= 15),
        ("multi_sport_athlete", sports_played >= 2),
        ("dedicated_player", total_games >= 50),
        ("elo_champion", max_elo >= 1500),
    ]
    
    for achievement_id, condition in achievement_checks:
        if achievement_id not in unlocked_achievement_ids and condition:
            # Unlock achievement
            try:
                user_achievement = UserAchievement(
                    user_id=user_id,
                    achievement_id=achievement_id,
                    unlocked_at=datetime.utcnow()
                )
                db.add(user_achievement)
                newly_unlocked.append(achievement_id)
            except Exception:
                # Skip if achievement doesn't exist or already unlocked
                continue
    
    if newly_unlocked:
        db.commit()
    
    # Get total stats
    total_unlocked = len(unlocked_achievement_ids) + len(newly_unlocked)
    total_points = db.query(func.sum(Achievement.points)).join(
        UserAchievement, Achievement.id == UserAchievement.achievement_id
    ).filter(UserAchievement.user_id == user_id).scalar() or 0
    
    return AchievementCheckResponse(
        newly_unlocked=newly_unlocked,
        total_unlocked=total_unlocked,
        total_points=total_points
    )

File: frontend-web/convex/achievements.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";
import { getAuthUserId } from "@convex-dev/auth/server";

export const list = query({
  args: {},
  handler: async (ctx) => {
    const all = await ctx.db.query("achievements").collect();
    return all.filter((a) => !a.isHidden);
  },
});

export const userAchievements = query({
  args: { userId: v.optional(v.id("users")) },
  handler: async (ctx, { userId }) => {
    if (!userId) return [];
    const unlocked = await ctx.db
      .query("userAchievements")
      .withIndex("by_user", (q) => q.eq("userId", userId))
      .collect();

    return Promise.all(
      unlocked.map(async (ua) => {
        const achievement = await ctx.db
          .query("achievements")
          .withIndex("by_achievement_id", (q) =>
            q.eq("achievementId", ua.achievementId),
          )
          .first();
        return {
          ...ua,
          achievement: achievement ?? {
            achievementId: ua.achievementId,
            name: ua.achievementId,
            description: "",
            category: "",
            points: 0,
            isHidden: false,
          },
        };
      }),
    );
  },
});

export const checkAndUnlock = mutation({
  args: {},
  handler: async (ctx) => {
    const userId = await getAuthUserId(ctx);
    if (!userId) throw new Error("Not authenticated");

    const existing = await ctx.db
      .query("userAchievements")
      .withIndex("by_user", (q) => q.eq("userId", userId))
      .collect();
    const unlockedSet = new Set(existing.map((e) => e.achievementId));

    // Get user stats
    const ratings = await ctx.db
      .query("userRatings")
      .withIndex("by_user_sport_mode", (q) => q.eq("userId", userId))
      .collect();

    const totalGames = ratings.reduce((s, r) => s + r.gamesPlayed, 0);
    const quizGames = ratings
      .filter((r) => r.mode === "quiz")
      .reduce((s, r) => s + r.gamesPlayed, 0);
    const survivalGames = ratings
      .filter((r) => r.mode === "survival")
      .reduce((s, r) => s + r.gamesPlayed, 0);
    const maxSurvivalScore = ratings
      .filter((r) => r.mode === "survival")
      .reduce((m, r) => Math.max(m, r.bestScore), 0);
    const maxElo = ratings.reduce((m, r) => Math.max(m, r.eloRating), 0);
    const sportsPlayed = new Set(ratings.map((r) => r.sport)).size;

    const user = await ctx.db.get(userId);
    const approvedQuestions = user?.approvedQuestionsCount ?? 0;

    const checks: [string, boolean][] = [
      ["first_quiz", quizGames >= 1],
      ["first_survival", survivalGames >= 1],
      ["survival_legend", maxSurvivalScore >= 15],
      ["multi_sport_athlete", sportsPlayed >= 2],
      ["dedicated_player", totalGames >= 50],
      ["elo_champion", maxElo >= 1500],
      ["the_architect", approvedQuestions >= 1],
    ];

    const newlyUnlocked: string[] = [];
    for (const [achId, met] of checks) {
      if (met && !unlockedSet.has(achId)) {
        await ctx.db.insert("userAchievements", {
          userId,
          achievementId: achId,
          unlockedAt: Date.now(),
        });
        newlyUnlocked.push(achId);
      }
    }

    return {
      newlyUnlocked,
      totalUnlocked: existing.length + newlyUnlocked.length,
    };
  },
});
